File: solutions/3987-maximum-weight-in-two-bags/maximum-weight-in-two-bags.py
class Solution:
    def maxWeight(self, weights: List[int], w1: int, w2: int) -> int:

        rows, cols = w1 + 2, w2 + 1

        nxt = cur = [[False]*(cols) for _ in range(rows)] 
        cur[0][0] = True
        
        for wgt in weights:
            cur, nxt = [tuple(row) for row in nxt], cur

            for tot1 in range(rows):
                for tot2 in range(cols):

                    if cur[tot1][tot2]:
                        if wgt <= w1 - tot1: 
                            nxt[tot1+wgt][tot2] = True
                        if wgt <= w2 - tot2: 
                            nxt[tot1][tot2+wgt] = True

            cur = nxt
            
        return max(tot1 + tot2 for tot1 in range(rows) 
                         for tot2 in range(cols) if cur[tot1][tot2])

File: solutions/3987-maximum-weight-in-two-bags/maximum-weight-in-two-bags.cpp
class Solution {
public:
    int function(int i, vector<int>& weights, vector<vector<vector<int>>>& dp, int w1, int w2) {
        if (i >= weights.size()){
            return 0;
        }

        if (dp[i][w1][w2] != -1){
            return dp[i][w1][w2];
        }

        int max_val = function(i+1,weights,dp,w1,w2);

        if (w1 >= weights[i]){
            max_val = max(max_val,weights[i] + function(i+1,weights,dp,w1-weights[i],w2));
        }

        if (w2 >= weights[i]){
            max_val = max(max_val,weights[i] + function(i+1,weights,dp,w1,w2-weights[i]));
        }

        return dp[i][w1][w2] = max_val;
    }

    int maxWeight(vector<int>& weights, int w1, int w2) {
        vector<vector<vector<int>>> dp = vector<vector<vector<int>>>(weights.size(),vector<vector<int>>(w1+1,vector<int>(w2+1,-1)));
        return function(0,weights,dp,w1,w2);
    }
};